### PR TITLE
OAuth social login/verification using Passport

### PR DESCRIPTION
**Pull request by tibetsprague**
Support Twitter, Github and Facebook for now.
Use a socket to communicate with client about when verification is complete.
Support local certificates to test SSL (needed to test Facebook login)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,6 +12,26 @@
...
+COOKIE_SECRET=anotherlongstringoftext
+```
+* If you want to support OAuth with Facebook, Twitter and/or GitHub you will need to add consumer client ids and secrets to .env
+```
+FACEBOOK_APP_ID={grab from app at https://developers.facebook.com/apps/214528186128969)
+FACEBOOK_APP_SECRET={grab from app at https://developers.facebook.com/apps/214528186128969}
+TWITTER_CONSUMER_KEY={grab from app at https://developer.twitter.com/en/apps/15974878}
+TWITTER_CONSUMER_SECRET={grab from app at https://developer.twitter.com/en/apps/15974878}
+GITHUB_CLIENT_ID={grab from https://github.com/settings/applications/944507}
+GITHUB_CLIENT_SECRET={grab from https://github.com/settings/applications/944507}
```
Comment (jellegerbrandy) [inline README.md]: These urls are only accessible if you have the correct permissions. Who is keeping the accounts? Can you give the creds to adam if you have no already done so?

File: common/models/account.js [hunk 2]
```diff
@@ -4,22 +4,26 @@ var sigUtil = require('eth-sig-util');
...
-      next(new Error('Invalid signature'));
-      return;
-    }
-
-    const text = "Please sign this message to confirm your request to update your profile to name '" + ctx.instance.name + "' and description '" + ctx.instance.description + "'. There's no gas cost to you. Timestamp:" + ctx.instance.timestamp;
-    const msg = ethUtil.bufferToHex(Buffer.from(text, 'utf8'));
-    const recoveredAddress = sigUtil.recoverPersonalSignature({ data: msg, sig: ctx.instance.signature });
-
-    if (recoveredAddress == ctx.instance.ethereumAccountAddress) {
+    if (ctx.options.skipSignatureCheck) {
```
Comment (jellegerbrandy) [inline common/models/account.js]: what is this for?

File: server/server.js [hunk 3]
```diff
@@ -1,29 +1,160 @@
...
+                   provider == 'facebook' ? profile._json.first_name + " " + profile._json.last_name : " ";
+      var account = await app.models.Account.findOne({ where: { ethereumAccountAddress: req.session.ethereumAccountAddress }});
+      if (!account) {
+        account = new app.models.Account();
+        account.ethereumAccountAddress = req.session.ethereumAccountAddress;
+        account.name = name;
+      }
+      account.userId = user.id;
+      account[provider + "URL"] = profileUrl;
+      account.save({ skipSignatureCheck: true });
```
Comment (jellegerbrandy) [inline server/server.js]: can you add a line explaining why we can skip the check?

File: server/server.js [hunk 4]
```diff
@@ -1,29 +1,160 @@
...
+    var options = {
+      key: privateKey,
+      cert: certificate
+    };
+    server = https.createServer(options, app);
+  } else {
+    server = http.createServer(app);
+  }
+  server.listen(app.get('port'), function() {
+    var baseUrl = (useHTTPS ? 'https://' : 'http://') + app.get('host') + ':' + app.get('port');
```
Comment (jellegerbrandy) [inline server/server.js]: This allowing of non-ssl connections is only for testing pruposes, I presume?

File: README.md (hunk 1 above)
Comment (tibetsprague) [reply README.md]: Yeah. Adam gave me the permissions so he already does. I think he needs to add you but I can try

File: common/models/account.js (hunk 2 above)
Comment (tibetsprague) [reply common/models/account.js]: We currently save the OAuth info the database all on the server side, and we don't have a signature from the client, so we need this to be able to save the database record

File: server/server.js (hunk 4 above)
Comment (tibetsprague) [reply server/server.js]: yes

**Pull request merged**